src/test/groovy/FindOldestStackVersionStepTests.groovy

```diff
@@ -0,0 +1,80 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import org.junit.Before
+import org.junit.Test
+import static org.junit.Assert.assertFalse
+import static org.junit.Assert.assertTrue
+
+class FindOldestStackVersionStepTests extends ApmBasePipelineTest {
+
+  @Override
+  @Before
+  void setUp() throws Exception {
+    super.setUp()
+    script = loadScript('vars/findOldestStackVersion.groovy')
+    helper.registerAllowedMethod('httpRequest', [Map.class], { f ->
+      return """{
+        "versions": [
+          "7.14.0-SNAPSHOT",
+          "7.14.0",
+          "7.14.1-SNAPSHOT",
+          "7.15.0-SNAPSHOT",
+          "7.15.0",
+          "8.0.0-SNAPSHOT"
+        ],
+        "aliases": [
+          "7.x-SNAPSHOT",
+          "7.14-SNAPSHOT",
+          "7.14",
+          "7.15-SNAPSHOT",
+          "7.15",
+          "8.0-SNAPSHOT",
+          "8.0"
+        ]
+      }"""
+    })
+  }
+
+  @Test
+  void test_missing_argument() throws Exception {
+    testMissingArgument('versionCondition') {
+      script.call()
+    }
+  }
+
+  @Test
+  void test_match() throws Exception {
+    def ret = script.call(versionCondition: "^7.14.0")
+    printCallStack()
+    assertTrue(ret.equals('7.14.0'))
+  }
+
+  @Test
+  void test_snapshot() throws Exception {
+    def ret = script.call(versionCondition: "^7.14.1")
+    printCallStack()
+    assertTrue(ret.equals('7.14.1-SNAPSHOT'))
+  }
+
```

Comment:
There is a corner case that happens during the FF of a release. There are artifacts for the version, the build candidates, but they haven't been released yet. In that case we want to use the snapshots, that are available. It'd be covered by a test case like this one:
```suggestion

  @Test
  void test_release_in_process() throws Exception {
    // There are already 7.15.0 artifacts, but this release hasn't happened yet, and there are only snapshots for the docker images.
    def ret = script.call(versionCondition: "^7.15.0")
    printCallStack()
    assertTrue(ret.equals('7.15.0-SNAPSHOT'))
  }
  
```

Reply:
I think that's already done in https://github.com/elastic/apm-pipeline-library/pull/1248/files#diff-b9e4d7001a12a17bcee07878bce19ae78c6e8a0e520d53d59db0e141b87d73e1R68 with the `artifacts-api/.../versions` data structure:

```
{
        "versions": [
          "7.14.0-SNAPSHOT",
          "7.14.0",
          "7.14.1-SNAPSHOT",
          "7.15.0-SNAPSHOT",
          "7.15.0",
          "8.0.0-SNAPSHOT"
        ],
        "aliases": [
          "7.x-SNAPSHOT",
          "7.14-SNAPSHOT",
          "7.14",
          "7.15-SNAPSHOT",
          "7.15",
          "8.0-SNAPSHOT",
          "8.0"
        ]
      }
```

Reply:
Notice that the cases are different:

* For `^7.14.0` there is a released version, but there are also snapshot artifacts, we want to test with the released version -> `7.14.0`.
* For `^7.14.1` we want to use `7.14.1-SNAPSHOT`, there are no `7.14.1` artifacts -> `7.14.1-SNAPSHOT`.
* For `^7.15.0` there is a released version and snapshot artifacts, but the version is not released yet and some artifacts as docker images are not available yet, we want to use the snapshots -> `7.15.0-SNAPSHOT`.

Reply:
See https://github.com/elastic/apm-pipeline-library/pull/1248/commits/f06532f631b5b195d0fd1eb1c0c9ff509b15795d

Thanks Jaime!